### ISSUE TITLE
New template.json file may need as template to generate random fake sample data.

### ISSUE THREAD

**Issue by ke-zhang-rd** [NSLS-II/amostra#41]
@danielballan 

We want generate random fake sample data but human readable, not random unicode code point. In hypothesis, people could achieve this by `alphabet` in `strategies.text`, doc [here](https://hypothesis.readthedocs.io/en/latest/data.html?highlight=alphabet#hypothesis.strategies.text).

We may want do same thing when call `from_schema`. Unfortunately or Fortunately, hypothesis-jsonschema doesn't allow people explicitly specify string alphabet. Alternatively, we could setup `pattern` in jsonschema. The role of current sample.json file is validation. A new file might be useful to separate template and validate roles.

**Comment by danielballan** [comment]
Does this talk offer any suggestions that could be helpful to us? https://youtu.be/jpDC8Sxcsbc

Offhand I think adding a `pattern` to our schemes sounds beneficial. Are there any downsides?

**Comment by ke-zhang-rd** [comment]
I will watch that talk sometime today. With a separating file, basically, this file will only be used internally, only in tests. It doesn't really influence any exposed performance. Which aspect are you thinking could be influenced?

Also another reason for having a separate file is `"required": ["revision", "uuid"]` is necessary in validation but illegal when create a new sample.

**Comment by ke-zhang-rd** [comment]
I just finished video. I found this [link](https://medium.com/capital-one-tech/why-you-dont-necessarily-need-data-for-data-science-48d7bf503074) from talk. 

It looks lots of people are working on generate synthetic data by ML from real data. Since we don't have real database, we instead specify `pattern` by human intelligence.

Should I create a `synthetic.py` under `amostra/amostra`?

**Comment by danielballan** [comment]
> Should I create a synthetic.py under amostra/amostra?

Sure.

> Also another reason for having a separate file is "required": ["revision", "uuid"] is necessary in validation but illegal when create a new sample.

Can you elaborate on that? Maybe the creation to should use `set_trait` so that it's legal? I think that having two copies of each schema---one for testing / example generation and one for validation---opens up the door for accidental divergence. It would be best to have _one_ schema and do our best to make it work for both, if at all practical.

**Comment by ke-zhang-rd** [comment]
> > Also another reason for having a separate file is "required": ["revision", "uuid"] is necessary in validation but illegal when create a new sample.
> 
> Can you elaborate on that? Maybe the creation should use set_trait so that it's legal?

Right. I assume `set_trait` is method behind `samples.new(kwargs->dict())`. We are talking about this dict(fake data) which is generated **before** feed to kwargs of `samples.new`. I agree, maybe use `set_trait` will make them auto insert `uuid` and `revision` legally later.

"required" field of jsonschema may make them diverge already. When validation is called, `"required"` need to be in jsonschema. When we generate sample, `"required"` need to be removed from jsonschema to let amostra handle `uuid` and `revision` insertion.

**Comment by ke-zhang-rd** [comment]
Is it necessary to make string value of fake sample data human readable?

Anyway, it's very expensive. The way Hypothesis yield string matching certain pattern in jsonschema is random generate a strings pool then check with pattern one by one. Then finally return those matched. It is possible that Hypothesis cannot find any and return message below

```
E   hypothesis.errors.FailedHealthCheck: It looks like your strategy is filtering out a lot of data. Health check found 50 filtered examples but only 0 good ones. This will make your tests much slower, and also will probably distort the data generation quite a lot. You should adapt your strategy to filter less. This can also be caused by a low max_leaves parameter in recursive() calls
E   See https://hypothesis.readthedocs.io/en/latest/healthchecks.html for more information about this. If you want to disable just this health check, add HealthCheck.filter_too_much to the suppress_health_check settings for this test.
```
There is way to money patch this [line](https://github.com/Zac-HD/hypothesis-jsonschema/blob/master/src/hypothesis_jsonschema/_impl.py#L615) by adding `alphabet` argument to fix it.

**Comment by danielballan** [comment]
I see your concern about required fields, but I think we can avoid creating two copies by solving this in a different way. Something like:

```py
# Use hypothesis to generate example sample (as dict).
sample_dict = generate_example_from_jsonschema(sample_schema)  # plain dict, includes revision, uuid
# Now create an amostra.objects.Sample object from this information.
# Handle the read-only traits specially.
revision = sample_dict.pop('revision')
uuid = sample_dict.pop('uuid')
sample = client.samples.new(**sample_dict)
sample.set_trait('revision', revision)
sample.set_trait('uuid', uuid)
```

To your second question, can you define "human readable"?

**Comment by ke-zhang-rd** [comment]
Ok, I will try to use one json file.
Now, I narrow down hypothesis random to english characters instead of unicode code point. When I said not human readable, I mean put no limitation on range of random unicode.

My opinion is we don't really need to make it human readable, cause we will compare expect and actual in code.
